ssl/ssl_sess.c

```diff
@@ -59,17 +59,37 @@ __owur static int timeoutcmp(SSL_SESSION *a, SSL_SESSION *b)
  */
 void ssl_session_calculate_timeout(SSL_SESSION *ss)
 {
-#ifndef __DJGPP__ /* time_t is unsigned on djgpp */
-    /* Force positive timeout */
-    if (ss->timeout < 0)
-        ss->timeout = 0;
+    time_t tmax = (time_t)-1;
+
+    if (sizeof(time_t) == 8) {
+        uint64_t overflow;
+
+#ifndef __DJGPP__ /* time_t is unsigned on djgpp, it's signed anywhere else */
+        tmax = (time_t)((uint64_t)tmax >> 1);
 #endif
-    ss->calc_timeout = ss->time + ss->timeout;
-    /*
-     * |timeout| is always zero or positive, so the check for
-     * overflow only needs to consider if |time| is positive
-     */
-    ss->timeout_ovf = ss->time > 0 && ss->calc_timeout < ss->time;
+        overflow = (uint64_t)tmax - (uint64_t)ss->time;
+        if (ss->timeout > (time_t)overflow) {
+            ss->timeout_ovf |= 1;
```

Comment:
Why `|= 1`? Why not just ss->timeout_ovf = 1? However the old code was also setting it to 0 if it was not overflowed. I do not really know if the session timeout values can change from overflowed to non-overflowed states, however it seems to me to be wrong to change the semantics.

Reply:
OK I will keep semantics.